src/Lean/Elab/Tactic/BVAckermannize.lean

```diff
@@ -0,0 +1,407 @@
+  | _root_.BitVec w => do
+     let w ← getNatValue? w
+     return .BitVec w
+  | _ => OptionT.fail
+
+/-- Convert a `BVTy` back into an `Expr` -/
+def toExpr : BVTy → Expr
+| .Bool => mkConst ``_root_.Bool
+| .BitVec w => mkApp (mkConst ``_root_.BitVec) (mkNatLit w)
+
+end BVTy
+
+/-- An argument to an uninterpreted function, which we track for ackermannization. -/
+structure Argument where
+  /-- The expression corresponding to the argument -/
+  x : Expr
+  /-- The cached type of the expression x -/
+  xTy : BVTy
+deriving Hashable, BEq, Inhabited
+
+namespace Argument
+
+instance : ToMessageData Argument where
+  toMessageData arg := m!"{arg.x} : {arg.xTy}"
+
+/-- Build an `Argument` from a raw expression of type `Bool` or `BitVec _`. -/
+def ofExpr? (e : Expr) : OptionT MetaM Argument := do
+  let t ← BVTy.ofExpr? (← inferType e)
+  return { x := e, xTy := t}
+
+end Argument
+
+/--
+A function, which packs the expression and the type of the codomain of the function.
+We use the type of the codomain to build an abstracted value for every call of this function.
+-/
+structure Function where
+  /-- The function -/
+  f : Expr
+  /-- The type of the function's codomain -/
+  codomain : BVTy
+ deriving Hashable, BEq, Inhabited
+namespace Function
+
+instance : ToMessageData Function where
+  toMessageData fn := m!"{fn.f} (cod: {fn.codomain})"
+
+/--
+Reify an expression `e` of the kind `f x₁ ... xₙ`, where all the arguments and the return type are
+a `BVTy` into an App.
+-/
+def reifyAp (f : Expr) : OptionT MetaM (Function × Array Argument) := do
+  let xs := f.getAppArgs
+  /- We need at least one argument for this to be a function call we can ackermannize. -/
+  guard <| xs.size > 0
+  let codomain ← BVTy.ofExpr? (← inferType f)
+  let args ← xs.mapM Argument.ofExpr?
+  let fn : Function := { f, codomain }
+  return (fn, args)
+end Function
+/--
+NOTE: is it sensible to hash an array of arguments?
+We may want to use something like a trie to index these.
+Consider swiching to something like `Trie`.
+-/
+abbrev ArgumentList := Array Argument
+/--
+The data stored for an ackermannized call to allow us to build proofs.
+-/
+structure CallVal where
+  /-- The free variable `ack_fx₁...xₙ := (f x₁ x₂ ... xₙ)`. -/
+  fvar : FVarId
+  /-- heqProof : The proof that `ack_fx₁...fxₙ = f x₁ x₂ ... xₙ` (name/fvar = value/expr). -/
+  heqProof : Expr
+deriving Hashable, BEq, Inhabited
+
+namespace CallVal
+
+instance : ToMessageData CallVal where
+  toMessageData cv := m!"{Expr.fvar cv.fvar} ({cv.heqProof})"
+
+end CallVal
+structure State where
+  /--
+  A mapping from a function `f`, to a map of arguments `x₁ ... xₙ`, to the information stored about the call.
+  This is used to generate equations of the form `x₁ = y₁ → x₂ = y₂ → ... → xₙ = yₙ → ack_fx₁...xₙ = ack_fy₁...yₙ on-demand.
+  -/
+  fn2args2call : Std.HashMap Function (Std.HashMap ArgumentList CallVal) := {}
+  /-- A counter for generating fresh names. -/
+  gensymCounter : Nat := 0
```

Comment:
Why can't we use the normal name generator for this? The hypotheses are going to be for machine consumption anyway so accessibility doesn't really matter.

Reply:
I prefer to generate nicer looking names :)